src/auth/authApp.js

```diff
@@ -84,23 +82,123 @@ export function redirectToSignInWithAuthRequest(authRequest: string = makeAuthRe
     return
   }
 
-  function successCallback() {
-    Logger.info('protocol handler detected')
-    // protocolCheck should open the link for us
+  // Create a unique ID used for this protocol detection attempt.
+  const echoReplyID = Math.random().toString(36).substr(2, 9)
+  const echoReplyKey = `echo-reply-${echoReplyID}`
+
+  // Use localStorage as a reliable cross-window communication method.
+  // Create the storage entry to signal a protocol detection attempt for the
+  // next browser window to check.
+  window.localStorage.setItem(echoReplyKey, 'pending')
+  const cleanUpLocalStorage = () => {
+    window.localStorage.removeItem(echoReplyKey)
   }
 
-  function failCallback() {
-    Logger.warn('protocol handler not detected')
-    window.location = httpsURI
+  const detectionTimeout = 1000
```

Comment:
Is this long enough?  What happens if I do this on an underpowered mobile phone?

Reply:
I've been calling these situations "false-negative" detections. On mobile a false redirect to https auth is largely inconsequential. The native mobile protocol handler app will have already launched.

Due to the nature of this new detection approach, these are unavoidable but we aim to make them unlikely and also not that bad when they do occur. 

See the initial PR description for more info
> False-positive detections are treated as unacceptable. Web app completely broken when user cannot login because we incorrectly assumed a protocol handler.
> False-negatives should be rare, but are acceptable. The web app may be redirect to https auth when the protocol handler worked.

Reply:
@kantai We detect when the "open blockstack app?" browser popup occurs, then pauses the redirect timeout. When the popup is closed (user clicks either yes or no/cancel), the timer is restarted. 

If the user clicks "yes open blockstack app", then the page visibility API detects that the protocol handler is *probably* working, and restarts the timer with 3000ms timeout. This is for the purpose of reducing probability of false-negative redirect, while still performing the redirect if something goes wrong with the protocol handler.

Reply:
Cool, that works for me, @zone117x -- and yeah, it correctly detected when that "open blockstack app?" popup occurs -- it might have been an older version I tested with that had that issue.